### PR TITLE
LPS-186928

### PR DESCRIPTION
**Pull request by marianoalvarosaiz**

### CODE REVIEW

**Comment by Kvale1733** [comment]
ci:test:clustering

File: modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java [hunk 1]
```diff
@@ -84,7 +88,8 @@
 	configurationPid = "com.liferay.portal.scheduler.internal.configuration.SchedulerEngineHelperConfiguration",
 	enabled = false, service = SchedulerEngineHelper.class
 )
-public class SchedulerEngineHelperImpl implements SchedulerEngineHelper {
+public class SchedulerEngineHelperImpl
+	implements ConfigurationModelListener, SchedulerEngineHelper {
```
Comment (dantewang) [inline modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java:92]: @marianoalvarosaiz This `ConfigurationModelListener` seems not needed, as you register an anonymous one in activate.

File: modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java [hunk 2]
```diff
@@ -642,6 +696,14 @@ public void removedService(
 						schedulerJobConfiguration.getName());
 
 			messageListenerServiceRegistration.unregister();
+
+			String factoryPid = (String)serviceReference.getProperty(
+				"factoryPid");
+
+			if (factoryPid != null) {
+				_factoryPidComponentConfigurations.put(
```
Comment (dantewang) [inline modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java:704]: @marianoalvarosaiz Should this be `remove`?

**Review by dantewang — COMMENTED** [review]
@marianoalvarosaiz The overall solution seems a bit hard to understand.

From the ticket, it seems there should be 2 jobs initialized by `UserImportSchedulerJobConfigurtion` for each company, but their names conflicts because the name provided by the `UserImportSchedulerJobConfigurtion` is the class name.

Is there other issues, like the configuration (for example `interval`) is not correctly picked up when creating the second job?

And why does adding a model listener to restart the component resolves this?

From my point of view, I would like to dynamically register the `UserImportSchedulerJobConfiguration` -- so it's not a component, but registered by some "Configurator", which handles the configuration changes and isolation.

What do you think?

@tinatian

File: modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java (hunk 1 above)
Comment (marianoalvarosaiz) [reply modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java:92]: @dantewang You're right. My mistake. Not needed. I left it unintentionally from a previous version.

**Comment by marianoalvarosaiz** [comment]
@dantewang 

> From the ticket, it seems there should be 2 jobs initialized by UserImportSchedulerJobConfigurtion for each company, but their names conflicts because the name provided by the UserImportSchedulerJobConfigurtion is the class name.

Not exactly. The LDAP Import in UserImportSchedulerJobConfigurtion is designed to depend on a common (system) interval, and then, during that interval, apply the specific company configuration.

> Is there other issues, like the configuration (for example interval) is not correctly picked up when creating the second job?

It isn't just a name collision. There are other issues. You cannot have a configurationPid for a factory like this because as soon as you save instance settings and system settings you'll get an error like:

```
2023-06-07 11:33:25.348 ERROR [CM Event Dispatcher (Fire ConfigurationEvent: pid=com.liferay.portal.security.ldap.exportimport.configuration.LDAPImportConfiguration)][LogService:56] Unexpected problem delivering configuration event to [org.osgi.service.cm.ConfigurationListener, id=79, bundle=61/file:/media/malvaro/datos1/dev/node1/osgi/static/org.apache.felix.configadmin.jar?protocol=jar&static=true]
java.lang.IllegalStateException: Factory pid supplied after all non-factory configurations supplied com.liferay.portal.security.ldap.exportimport.configuration.LDAPImportConfiguration
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.checkFactoryPidIndex(ConfigurableComponentHolder.java:535) ~[?:?]
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.configurationUpdated(ConfigurableComponentHolder.java:366) ~[?:?]
        at org.apache.felix.scr.impl.manager.RegionConfigurationSupport.configureComponentHolder(RegionConfigurationSupport.java:191) ~[?:?]
        at org.apache.felix.scr.impl.manager.RegionConfigurationSupport.configurationEvent(RegionConfigurationSupport.java:286) ~[?:?]
        at org.apache.felix.scr.impl.manager.RegionConfigurationSupport$2.configurationEvent(RegionConfigurationSupport.java:119) ~[?:?]
        at org.apache.felix.cm.impl.ConfigurationManager$FireConfigurationEvent.sendEvent(ConfigurationManager.java:1686) [bundleFile:?]
        at org.apache.felix.cm.impl.ConfigurationManager$FireConfigurationEvent.run(ConfigurationManager.java:1637) [bundleFile:?]
```

> And why does adding a model listener to restart the component resolves this?
The configuration model listener is what we had before to overcome this problem https://github.com/liferay/liferay-portal/blob/7.3.x/modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/configuration/persistence/listener/UserImportConfigurationModelListener.java

With the refresh inside the scheduler class I avoid having to change back the interface to allow the register/unregister so that they can be called each time the configuration is modified. Also the listener allows to discriminate the configuration we want, the sytem one.


For CheckArticleSchedulerJobConfiguration is even worse because it isn't notified of the configuration change.

**Comment by marianoalvarosaiz** [comment]
@dantewang I've sent a new pull with your considerations, see https://github.com/liferay-core-infra/liferay-portal/pull/4777

**Pull request closed without merging**